source/extensions/common/wasm/wasm.cc

```diff
@@ -38,9 +15,22 @@ namespace Wasm {
 
 namespace {
 
+struct CodeCacheEntry {
+  std::string code;
+  bool in_progress;
+  MonotonicTime fetch_time;
+};
+
 const std::string INLINE_STRING = "<inline>";
+// NB: xDS currently does not support failing asynchronously, so we fail immediately
+// if remote Wasm code is not cached and do a background fill.
+const bool DEFAULT_FAIL_IF_NOT_CACHED = true;
+bool fail_if_code_not_cached = DEFAULT_FAIL_IF_NOT_CACHED;
+const int CODE_CACHE_SECONDS_NEGATIVE_CACHING = 10;
 
-std::atomic<int64_t> active_wasm_;
+std::atomic<int64_t> active_wasms;
+std::mutex code_cache_mutex;
+std::unordered_map<std::string, CodeCacheEntry>* code_cache = nullptr;
```

Comment:
This is an unbounded cache. is that ok? What is the lifetime?  server or listener,filter?

Reply:
The lifetime is unbounded.  SInce it is SHA based I felt like that was safe.